site/content/coursework2.md

```diff
@@ -0,0 +1,424 @@
+    1. Neumann conditions $\nabla u \cdot n = h$?
+1. Would this spatial discretisation have the same timestep restriction
+   as the 2nd order operator, or a different one? Explain your answer.
+
+{{< /question >}}
+
+## Part 3: multigrid
+
+### Part 3a: completing the multigrid solver
+Implement the missing pieces in the `MGSolver` class, namely
+
+- `MGSolver.jacobi`
+- `MGSolver.vcycle`
+- `MGSolver.wcycle`
+
+You should do this directly in the `mgsolver/mgsolver.py` file (don't
+forget to commit it!). If you do this correctly, the one dimensional
+tests should now pass.
+
+{{< hint info >}}
+Ensure that your implementation is correct in both serial and
+parallel. Up to round-off error, you should get the same results
+independent of the number of processes.
+{{< /hint >}}
+
+### Part 3b: solving for a steady state
+
+{{< hint info >}}
+You should do your implementation for this part in a file called
+`part2_multigrid.py` placed in the root of the repository.
+{{< /hint >}}
+
+Using the same `Poisson7pt` operator that you implemented for Part
+1a, we will now solve for the steady state directly (rather than
+timestepping towards it).
+
+Confirm that your implementation is correct by doing an MMS
+convergence test. For large problems you will probably want to run in
+parallel.
+
+{{< hint info >}}
+If your operator definition was correct in parallel in Part 1, you
+should not have to worry very hard about parallelism in this part,
+since everything is done with "collective" operations.
+{{< /hint >}}
+
+{{< question "Part 2b questions" >}}
+
+1. What mesh convergence do you get for this problem? Do you have to
+   adjust the tolerance to which you solve the problem as you add more
+   grid levels?
+1. For this problem, which method (jacobi, V-cycles, W-cycles) works
+   best when you add more grid levels?
+
+   Consider both algorithmic convergence and time to solution.
+1. Play around with the number of smoothing steps, does that change
+   your conclusions?
+
+{{< details "Performance hint" >}}
+
+You may wish to pull the body of the matrix-vector `mult` method out
+and try JIT-compiling it with numba. I found that made a big
+(positive) difference in the performance of my code.
+
+{{< /details >}}
+
+{{< /question >}}
+
+
+## Submission and mark scheme {#submission}
+
+The work will be marked on the basis of three things
+
+1. Your submitted code;
+2. A short report discussing answers to the questions and your
+   findings;
+3. A brief (10 min) oral exam with the lecturers. We will use this to
+   have a brief discussion about your implementation choices and code,
+   and any interesting things you found in your numerical experiments.
+   No need to prepare anything specific.
+
+You should submit to DUO a zip file containing
+
+1. A PDF of your writeup (max 4 pages), use your Z-code to name this
+   as ZCODE.pdf;
+2. A text file ZCODE.txt containing the commit hash of the code on
+   github you want us to mark.

```

Comment:
Submission on ULTRA should just be commit hash, and the writeup should live in the repository, I think.

Reply:
changed